torchaudio/datasets/musdb_hq.py

```diff
@@ -0,0 +1,103 @@
+import os
+from pathlib import Path
+from typing import Tuple, Union, List, Optional
+
+import torch
+import torchaudio
+from torch.utils.data import Dataset
+
+_URL = "https://zenodo.org/record/3338373/files/musdb18hq.zip"
+_EXT = ".wav"
+_SAMPLE_RATE = 44100
+
+
+class MUSDB_HQ(Dataset):
+    """Create *MUSDB_HQ Dataset* [:footcite:`MUSDB18HQ`] Dataset
+
+    Args:
+        root (str or Path): Root directory where the dataset's top level directory is found
+        subset (str): Subset of the dataset to use. Options: [``"train"``, ``"test"``].
+        sources (List[str] or None, optional): Sources extract data from.
+            Options: [``bass``, ``drums``, ``other``, ``mixture``, ``vocals``].
+            If ``None``, dataset consists of tracks except mixture.
+            (default: ``"[``bass``, ``drums``, ``other``, ``vocals``]"``)
+        split (str, optional): Whether to split training set into train and validation set.
+            If None, no splitting occurs. If ``train`` or ``validation``, returns respective set.
+            (default: ``None``)
+        validation (List[str] or None, optional): Tracks which are included in validation set.
+            If ``None``, dataset consists of no tracks in validation set.
+            (default: ``None``)
+    """
+
+    def __init__(
+        self,
+        root: Union[str, Path],
+        subset: str,
+        sources: List[str] = ["bass", "drums", "other", "vocals"],
+        split: Optional[str] = None,
+        validation: Optional[List[str]] = None,
+    ) -> None:
+        self.sources = sources
+        self.split = split
+        self.validation = validation
+
+        basename = os.path.basename(_URL)
+        basename = basename.rsplit(".", 2)[0]
+        assert subset in ["test", "train"], "`subset` must be one of ['test', 'train']"
+        assert self.split is None or self.split in [
+            "train",
+            "validation",
+        ], "`split` must be one of ['train', 'validation']"
+        assert (
+            self.split is None and self.validation is None or self.split and self.validation
+        ), "`split` and `validation` must both be None or not be None"
+        self._path = os.path.join(root, basename, subset)
+        self.names = self._collect_songs()
+
+    def _get_track(self, name, source):
+        return Path(self._path) / name / f"{source}{_EXT}"
+
+    def _load_sample(self, n: int) -> Tuple[torch.Tensor, int, int, str]:
+        name = self.names[n]
+        wavs = []
+        sample_rate = None
+        num_frames = None
+        for source in self.sources:
+            track = self._get_track(name, source)
+            wav, _ = torchaudio.load(str(track))
+            if sample_rate is None:
```

Comment:
my suggestion, since `sample_rate` should always match `_SAMPLE_RATE`, would be to do something along the lines of

```
    wav, sr = torchaudio.load(str(track))
    assert sr == _SAMPLE_RATE, f"expected sample rate {_SAMPLE_RATE}, but got {sr}"
    if num_frames is None:
         num_frames = info.num_frames
    else:
         assert info.num_frames == num_frames, "num_frames do not match across sources"
    ...
    return stacked, _SAMPLE_RATE, num_frames, name
```